File: tools/devops-mcp/src/devops_mcp/providers/azdo.py
"""Azure DevOps Services (dev.azure.com) Repos + Pipelines via the REST API (7.1)."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

import httpx

from ..auth import AdoAuth
from ..errors import DevOpsError
from ..repo import Remote
from .base import (
    Check,
    FailedStep,
    Http,
    JobResult,
    PipelineRun,
    Provider,
    PullRequest,
    RunReport,
    tail,
)

API_VERSION = "7.1"

_RESULTS = {
    "succeeded": "success",
    "succeededWithIssues": "partial",
    "partiallySucceeded": "partial",
    "failed": "failure",
    "canceled": "cancelled",
    "abandoned": "cancelled",
    "skipped": "skipped",
    "none": None,
}

_POLICY = {
    "approved": ("completed", "success"),
    "rejected": ("completed", "failure"),
    "broken": ("completed", "failure"),
    "notApplicable": ("completed", "skipped"),
    "running": ("in_progress", None),
    "queued": ("queued", None),
}

_PR_STATUS = {
    "succeeded": ("completed", "success"),
    "failed": ("completed", "failure"),
    "error": ("completed", "failure"),
    "notApplicable": ("completed", "skipped"),
    "pending": ("in_progress", None),
    "notSet": ("queued", None),
}


def _result(value: str | None) -> str | None:
    return _RESULTS.get(value, value) if value else None


def _status(value: str | None) -> str:
    if value == "completed":
        return "completed"
    if value in ("inProgress", "cancelling"):
        return "in_progress"
    return "queued"


def _short(ref: str | None) -> str:
    return (ref or "").removeprefix("refs/heads/")


class AzureDevOpsProvider(Provider):
    kind = "Azure DevOps"
    # ADO rejects pull request descriptions longer than 4000 characters.
    max_body_length = 4000

    def __init__(self, remote: Remote, client: httpx.AsyncClient | None = None) -> None:
        self.remote = remote
        self.org_url = f"https://dev.azure.com/{quote(remote.owner, safe='')}"
        self.project_url = f"{self.org_url}/{quote(remote.project or '', safe='')}"
        self.repo_url = f"{self.project_url}/_git/{quote(remote.repo, safe='')}"
        self.http = Http(AdoAuth(), {"Accept": "application/json"}, client)
        self._repo: dict[str, Any] | None = None

    async def _api(self, method: str, path: str, params: dict[str, Any] | None = None, **kwargs: Any) -> Any:
        query = {"api-version": API_VERSION, **(params or {})}
        return await self.http.json(method, f"{self.project_url}/_apis/{path}", params=query, **kwargs)

    async def _repo_info(self) -> dict[str, Any]:
        if self._repo is None:
            self._repo = await self._api("GET", f"git/repositories/{quote(self.remote.repo, safe='')}")
        return self._repo

    async def _pr_path(self, suffix: str = "") -> str:
        repo = await self._repo_info()
        return f"git/repositories/{repo['id']}/pullrequests{suffix}"

    def _pr(self, d: dict[str, Any]) -> PullRequest:
        state = {"active": "open", "completed": "merged", "abandoned": "closed"}.get(d.get("status", ""), d.get("status", ""))
        return PullRequest(
            id=d["pullRequestId"],
            title=d.get("title", ""),
            body=d.get("description") or "",
            state=state,
            draft=bool(d.get("isDraft")),
            source_branch=_short(d.get("sourceRefName")),
            target_branch=_short(d.get("targetRefName")),
            author=(d.get("createdBy") or {}).get("displayName", ""),
            url=f"{self.repo_url}/pullrequest/{d['pullRequestId']}",
            head_sha=(d.get("lastMergeSourceCommit") or {}).get("commitId"),
        )

    def _run(self, d: dict[str, Any]) -> PipelineRun:
        return PipelineRun(
            id=d["id"],
            name=f"{(d.get('definition') or {}).get('name', 'build')} #{d.get('buildNumber', d['id'])}",
            status=_status(d.get("status")),
            result=_result(d.get("result")) if d.get("status") == "completed" else None,
            branch=_short(d.get("sourceBranch")),
            commit=d.get("sourceVersion"),
            url=((d.get("_links") or {}).get("web") or {}).get("href")
            or f"{self.project_url}/_build/results?buildId={d['id']}",
            started=d.get("startTime") or d.get("queueTime"),
            finished=d.get("finishTime"),
            event=d.get("reason"),
        )

    async def default_branch(self) -> str:
        branch = _short((await self._repo_info()).get("defaultBranch"))
        if not branch:
            raise DevOpsError("The Azure DevOps repository has no default branch yet.")
        return branch

    async def find_open_pr(self, branch: str) -> PullRequest | None:
        data = await self._api(
            "GET",
            await self._pr_path(),
            {"searchCriteria.sourceRefName": f"refs/heads/{branch}", "searchCriteria.status": "active"},
        )
        prs = data.get("value", [])
        return self._pr(prs[0]) if prs else None

    async def get_pr(self, pr_id: int) -> PullRequest:
        return self._pr(await self._api("GET", await self._pr_path(f"/{pr_id}")))

    async def create_pr(self, title: str, body: str, source: str, target: str, draft: bool) -> PullRequest:
        payload = {
            "sourceRefName": f"refs/heads/{source}",
            "targetRefName": f"refs/heads/{target}",
            "title": title,
            "description": body,
            "isDraft": draft,
        }
        return self._pr(await self._api("POST", await self._pr_path(), json=payload))

    async def update_pr(self, pr_id: int, title: str | None, body: str | None) -> PullRequest:
        payload = {k: v for k, v in (("title", title), ("description", body)) if v is not None}
        return self._pr(await self._api("PATCH", await self._pr_path(f"/{pr_id}"), json=payload))

    async def list_runs(self, branch: str | None, pr: PullRequest | None, limit: int) -> list[PipelineRun]:
        repo = await self._repo_info()
        params: dict[str, Any] = {
            "$top": limit,
            "queryOrder": "queueTimeDescending",
            "repositoryId": repo["id"],
            "repositoryType": "TfsGit",
        }
        if pr is not None:
            params["branchName"] = f"refs/pull/{pr.id}/merge"
        elif branch:
            params["branchName"] = f"refs/heads/{branch}"
        data = await self._api("GET", "build/builds", params)
        return [self._run(b) for b in data.get("value", [])]

    async def run_report(self, run_id: int, log_lines: int) -> RunReport:
        run = self._run(await self._api("GET", f"build/builds/{run_id}"))
        timeline = await self._api("GET", f"build/builds/{run_id}/timeline") or {}
        records = sorted(timeline.get("records") or [], key=lambda r: (r.get("order") or 0))
        by_id = {r["id"]: r for r in records}

        def job_of(rec: dict[str, Any]) -> str:
            cur: dict[str, Any] | None = rec
            while cur is not None and cur.get("type") != "Job":
                cur = by_id.get(cur.get("parentId"))
            return cur["name"] if cur else "(unknown job)"

        jobs = [
            JobResult(r["name"], _status(r.get("state")), _result(r.get("result")), f"{run.url}&view=logs&j={r['id']}")
            for r in records
            if r.get("type") == "Job"
        ]
        failures: list[FailedStep] = []
        for r in records:
            if r.get("type") != "Task" or r.get("result") != "failed":
                continue
            errors = [i["message"] for i in r.get("issues") or [] if i.get("type") == "error"][:20]
            log = await self._log(run_id, (r.get("log") or {}).get("id"), log_lines)
            failures.append(FailedStep(job_of(r), r["name"], errors, log))
        return RunReport(run, jobs, failures)

    async def _log(self, run_id: int, log_id: int | None, lines: int) -> str | None:
        if log_id is None or lines <= 0:
            return None
        try:
            resp = await self.http.request(
                "GET",
                f"{self.project_url}/_apis/build/builds/{run_id}/logs/{log_id}",
                params={"api-version": API_VERSION},
                headers={"Accept": "text/plain"},
            )
        except DevOpsError:
            return None
        return tail(resp.text, lines)

    async def pr_checks(self, pr: PullRequest) -> list[Check]:
        repo = await self._repo_info()
        artifact = f"vstfs:///CodeReview/CodeReviewId/{repo['project']['id']}/{pr.id}"
        data = await self._api("GET", "policy/evaluations", {"artifactId": artifact, "api-version": "7.1-preview.1"})
        checks: list[Check] = []
        for ev in data.get("value", []):
            cfg = ev.get("configuration") or {}
            name = (cfg.get("settings") or {}).get("displayName") or (cfg.get("type") or {}).get("displayName", "policy")
            status, result = _POLICY.get(ev.get("status", ""), ("queued", None))
            build_id = (ev.get("context") or {}).get("buildId")
            url = f"{self.project_url}/_build/results?buildId={build_id}" if build_id else None
            checks.append(Check(name, status, result, bool(cfg.get("isBlocking")), url))

        # External CI systems report through PR statuses; keep only the newest per context.
        statuses = (await self._api("GET", await self._pr_path(f"/{pr.id}/statuses"))).get("value", [])
        latest: dict[str, dict[str, Any]] = {}
        for s in sorted(statuses, key=lambda s: s.get("id", 0)):
            ctx = s.get("context") or {}
            latest["/".join(p for p in (ctx.get("genre"), ctx.get("name")) if p)] = s
        for name, s in latest.items():
            status, result = _PR_STATUS.get(s.get("state", ""), ("queued", None))
            checks.append(Check(name, status, result, None, s.get("targetUrl")))
        return checks

File: apps/vscode/src/services/devops-mcp/server/providers/azdo.ts
/** Azure DevOps Services (dev.azure.com) Repos + Pipelines through the REST API (7.1). */
import { type Auth, adoAuth } from "../auth"
import { DevOpsError } from "../errors"
import type { Remote } from "../repo"
import {
	type Check,
	type FailedStep,
	type Fetch,
	Http,
	type JobResult,
	type PipelineRun,
	type Provider,
	type PullRequest,
	type Result,
	type RunReport,
	type Status,
	tail,
} from "./types"

const API_VERSION = "7.1"

const RESULTS: Record<string, string | undefined> = {
	succeeded: "success",
	succeededWithIssues: "partial",
	partiallySucceeded: "partial",
	failed: "failure",
	canceled: "cancelled",
	abandoned: "cancelled",
	skipped: "skipped",
	none: undefined,
}

const POLICY: Record<string, [Status, Result]> = {
	approved: ["completed", "success"],
	rejected: ["completed", "failure"],
	broken: ["completed", "failure"],
	notApplicable: ["completed", "skipped"],
	running: ["in_progress", undefined],
	queued: ["queued", undefined],
}

const PR_STATUS: Record<string, [Status, Result]> = {
	succeeded: ["completed", "success"],
	failed: ["completed", "failure"],
	error: ["completed", "failure"],
	notApplicable: ["completed", "skipped"],
	pending: ["in_progress", undefined],
	notSet: ["queued", undefined],
}

const result = (value?: string | null): Result => (value ? (value in RESULTS ? RESULTS[value] : value) : undefined)

const status = (value?: string | null): Status =>
	value === "completed" ? "completed" : value === "inProgress" || value === "cancelling" ? "in_progress" : "queued"

const short = (ref?: string | null) => (ref ?? "").replace(/^refs\/heads\//, "")

export class AzureDevOpsProvider implements Provider {
	readonly kind = "Azure DevOps"
	// Azure DevOps rejects pull request descriptions longer than 4000 characters.
	readonly maxBodyLength = 4000
	readonly repoUrl: string
	private readonly projectUrl: string
	private readonly http: Http
	private repoInfo?: Promise<any>

	constructor(
		private readonly remote: Remote,
		fetchImpl?: Fetch,
		auth: Auth = adoAuth(),
	) {
		const orgUrl = `https://dev.azure.com/${encodeURIComponent(remote.owner)}`
		this.projectUrl = `${orgUrl}/${encodeURIComponent(remote.project ?? "")}`
		this.repoUrl = `${this.projectUrl}/_git/${encodeURIComponent(remote.repo)}`
		this.http = new Http(auth, { Accept: "application/json" }, fetchImpl)
	}

	private api<T = any>(method: string, path: string, query: Record<string, string | number | undefined> = {}, json?: unknown) {
		return this.http.json<T>(method, `${this.projectUrl}/_apis/${path}`, {
			query: { "api-version": API_VERSION, ...query },
			json,
		})
	}

	private repo(): Promise<any> {
		if (!this.repoInfo) {
			this.repoInfo = this.api("GET", `git/repositories/${encodeURIComponent(this.remote.repo)}`)
			// Do not cache a failure (e.g. before the user has signed in).
			this.repoInfo.catch(() => {
				this.repoInfo = undefined
			})
		}
		return this.repoInfo
	}

	private async prPath(suffix = ""): Promise<string> {
		return `git/repositories/${(await this.repo()).id}/pullrequests${suffix}`
	}

	private pr(d: any): PullRequest {
		const states: Record<string, string> = { active: "open", completed: "merged", abandoned: "closed" }
		return {
			id: d.pullRequestId,
			title: d.title ?? "",
			body: d.description ?? "",
			state: states[d.status] ?? d.status ?? "",
			draft: Boolean(d.isDraft),
			sourceBranch: short(d.sourceRefName),
			targetBranch: short(d.targetRefName),
			author: d.createdBy?.displayName ?? "",
			url: `${this.repoUrl}/pullrequest/${d.pullRequestId}`,
			headSha: d.lastMergeSourceCommit?.commitId,
		}
	}

	private run(d: any): PipelineRun {
		return {
			id: d.id,
			name: `${d.definition?.name ?? "build"} #${d.buildNumber ?? d.id}`,
			status: status(d.status),
			result: d.status === "completed" ? result(d.result) : undefined,
			branch: short(d.sourceBranch),
			commit: d.sourceVersion,
			url: d._links?.web?.href ?? `${this.projectUrl}/_build/results?buildId=${d.id}`,
			started: d.startTime ?? d.queueTime,
			finished: d.finishTime,
			event: d.reason,
		}
	}

	async defaultBranch(): Promise<string> {
		const branch = short((await this.repo()).defaultBranch)
		if (!branch) {
			throw new DevOpsError("The Azure DevOps repository has no default branch yet.")
		}
		return branch
	}

	async findOpenPr(branch: string): Promise<PullRequest | undefined> {
		const data = await this.api("GET", await this.prPath(), {
			"searchCriteria.sourceRefName": `refs/heads/${branch}`,
			"searchCriteria.status": "active",
		})
		const prs: any[] = data.value ?? []
		return prs.length ? this.pr(prs[0]) : undefined
	}

	async getPr(id: number): Promise<PullRequest> {
		return this.pr(await this.api("GET", await this.prPath(`/${id}`)))
	}

	async createPr(title: string, body: string, source: string, target: string, draft: boolean): Promise<PullRequest> {
		const json = {
			sourceRefName: `refs/heads/${source}`,
			targetRefName: `refs/heads/${target}`,
			title,
			description: body,
			isDraft: draft,
		}
		return this.pr(await this.api("POST", await this.prPath(), {}, json))
	}

	async updatePr(id: number, title: string | undefined, body: string | undefined): Promise<PullRequest> {
		const json: Record<string, string> = {}
		if (title !== undefined) json.title = title
		if (body !== undefined) json.description = body
		return this.pr(await this.api("PATCH", await this.prPath(`/${id}`), {}, json))
	}

	async listRuns(branch: string | undefined, pr: PullRequest | undefined, limit: number): Promise<PipelineRun[]> {
		const repo = await this.repo()
		const query: Record<string, string | number | undefined> = {
			$top: limit,
			queryOrder: "queueTimeDescending",
			repositoryId: repo.id,
			repositoryType: "TfsGit",
		}
		if (pr) {
			query.branchName = `refs/pull/${pr.id}/merge`
		} else if (branch) {
			query.branchName = `refs/heads/${branch}`
		}
		const data = await this.api("GET", "build/builds", query)
		return (data.value ?? []).map((b: any) => this.run(b))
	}

	async runReport(runId: number, logLines: number): Promise<RunReport> {
		const run = this.run(await this.api("GET", `build/builds/${runId}`))
		const timeline = (await this.api("GET", `build/builds/${runId}/timeline`)) ?? {}
		const records: any[] = [...(timeline.records ?? [])].sort((a, b) => (a.order ?? 0) - (b.order ?? 0))
		const byId = new Map(records.map((r) => [r.id, r]))
		const jobOf = (rec: any): string => {
			let cur = rec
			while (cur && cur.type !== "Job") {
				cur = byId.get(cur.parentId)
			}
			return cur?.name ?? "(unknown job)"
		}

		const jobs: JobResult[] = records
			.filter((r) => r.type === "Job")
			.map((r) => ({
				name: r.name,
				status: status(r.state),
				result: result(r.result),
				url: `${run.url}&view=logs&j=${r.id}`,
			}))
		const failures: FailedStep[] = []
		for (const r of records) {
			if (r.type !== "Task" || r.result !== "failed") {
				continue
			}
			const errors = (r.issues ?? [])
				.filter((i: any) => i.type === "error")
				.map((i: any) => i.message)
				.slice(0, 20)
			failures.push({ job: jobOf(r), step: r.name, errors, logTail: await this.log(runId, r.log?.id, logLines) })
		}
		return { run, jobs, failures }
	}

	private async log(runId: number, logId: number | undefined, lines: number): Promise<string | undefined> {
		if (logId === undefined || lines <= 0) {
			return undefined
		}
		try {
			const resp = await this.http.request("GET", `${this.projectUrl}/_apis/build/builds/${runId}/logs/${logId}`, {
				query: { "api-version": API_VERSION },
				accept: "text/plain",
			})
			return tail(await resp.text(), lines)
		} catch (error) {
			if (error instanceof DevOpsError) return undefined
			throw error
		}
	}

	async prChecks(pr: PullRequest): Promise<Check[]> {
		const repo = await this.repo()
		const artifactId = `vstfs:///CodeReview/CodeReviewId/${repo.project.id}/${pr.id}`
		const evaluations = await this.api("GET", "policy/evaluations", { artifactId, "api-version": "7.1-preview.1" })
		const checks: Check[] = (evaluations.value ?? []).map((ev: any) => {
			const cfg = ev.configuration ?? {}
			const [evStatus, evResult] = POLICY[ev.status] ?? ["queued", undefined]
			const buildId = ev.context?.buildId
			return {
				name: cfg.settings?.displayName ?? cfg.type?.displayName ?? "policy",
				status: evStatus,
				result: evResult,
				required: Boolean(cfg.isBlocking),
				url: buildId ? `${this.projectUrl}/_build/results?buildId=${buildId}` : undefined,
			}
		})

		// External CI systems report through PR statuses; keep only the newest per context.
		const statuses: any[] = (await this.api("GET", await this.prPath(`/${pr.id}/statuses`))).value ?? []
		const latest = new Map<string, any>()
		for (const s of [...statuses].sort((a, b) => (a.id ?? 0) - (b.id ?? 0))) {
			const name = [s.context?.genre, s.context?.name].filter(Boolean).join("/")
			latest.set(name, s)
		}
		for (const [name, s] of latest) {
			const [sStatus, sResult] = PR_STATUS[s.state] ?? ["queued", undefined]
			checks.push({ name, status: sStatus, result: sResult, url: s.targetUrl })
		}
		return checks
	}
}
